### ISSUE TITLE
[Design] Custom Tokens

### ISSUE THREAD

**Issue by r4mmer** [HathorNetwork/hathor-ledger-app#28]
# Summary
[summary]: #summary

Add support for custom tokens on Ledger.

# Motivation
[motivation]: #motivation

Many users send custom tokens and NFTs to a Ledger generated address and later found out they can't send the custom tokens and NFTs out of the Ledger.
To make transactions with custom tokens and NFTs, users need a software wallet with the same seed as the Ledger, this defeats the purpose of using the Ledger in the first place.

This feature would enable many users to send any custom tokens from the Ledger.

A custom token is identified by it's unique id, but users usually use the symbol or name which are not unique.
This introduces a security problem on the Ledger UX if we were to show the user only the symbol (e.g. ABC) when validating a transaction.
A possible attacker may create another token with the same symbol and the user will not have a way to verify that the token is actually correct.
The only secure option would be recognizing the unique id, but it's a 64 character long hex string which is very user-unfriendly.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

To ensure the safety of the funds with a more user-friendly approach the token will be confirmed using all information available, specially the unique id.
If the user recognizes the token, these available information will be signed by the Ledger (token signature) and saved on the wallet to use later.

The process to send a transaction without a custom token should remain unchanged, but we have to add some steps when dealing with custom tokens.
When a wallet is trying to sign a tx with a custom token it should send all token information (along with token signatures) before sending the transaction.
If any tokens are not confirmed or the signatures are invalid, the wallet must make sure they are confirmed before the transaction can be sent.

The sign_tx command will show token symbol and amount separatedly so the user can confirm the symbol matches.
Verifying only the symbol here is not a problem since the signature provided makes sure the user has already confirmed the unique id.

Effectively this makes a Ledger confirmation and signature counterpart of registered tokens.
This means we could ask the user for confirmation on the Ledger when registering the token, the signature will be made without the need to send a tx and with minimal changes to the UI.

We should also consider the case of a user wanting to "untrust" a token.
If the signature information on the wallet is leaked it could be used as a posible attack vector and the Ledger would have no way of knowing that the user has "untrusted" the token.
This is why we need to save a secret on the persistent storage to be included on the signature.
This secret can be generated with `cx_rng` since this method uses Ledger native RNG it will not be easily guessed. 

An "untrust" command can be issued at any time, the handler will generate and save another secret using the same method.
All signatures made before the command was issued will be nulled since the verification method will use another salt.

The wallet should save all signatures along with the token information, but allowing tokens to be registered even without ledger signatures since the ledger confirmation can be done at any time.

There should also be a command to check a signature so the wallet can check that the registered tokens are still valid.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

We will be implementing 4 new apdu commands and adding some functionality to the existing `sign_tx` command.
The permanent storage of Ledger will be used to store a secret of 32 bytes (`uint8_t secret[32]`) to work as salt.

Token symbol and name are UTF-8 encoded but the ledger cannot display UTF-8 characters.
We could remove or change these characters to a box character but it still wouldn't solve the issue. 
Not showing part of the symbol or name could be very confusing in some cases (e.g. the symbol is in japanese) or even a security concern.
To avoid these cases we are only allowing ascii only symbols and names.

### SIGN_TOKEN_DATA

This new command will receive the token information below and upon user confirmation it will sign the information and return the signature.

Token information needed to sign:
- Token UID: lenth == 32 bytes
- Token Symbol: length <= 5 (\*chars)
- Token Name: length <= 30 (\*chars)
- Token Version: Should always be 1 (encoded in 1 byte)

\* Obs: name and symbol are encoded as UTF-8 which uses 1 to 4 bytes of data for each character.
        Even if we reject the token with non-ascii characters, we should consider the length of the message for all tokens.

These informations will be encoded to be passed on the APDU command as:
```
token_uid[32] + len(symbol)[1] + symbol + len(name)[1] + name + version[1]
```
This will give us a max of 175 bytes (using max length and when UTF-8 uses 4 bytes, 70 when using 1 byte)
Since an APDU command let's us pass up to 255 bytes of data we can always encode this command on 1 call.

The screen to confirm should have the steps:
- Display token symbol
- Display token name
- Display token UID
- Step with confirm action
- Step with reject action

Upon confirmation the data will be signed according to the method described on the [methods](#methods) section.

### VERIFY_TOKEN_SIGNATURE

This new command will receive the token information and a signature to verify.
It will (without user confirmation) return if the signature is valid.

Token information needed:
- Token UID: lenth == 32 bytes
- Token Symbol: length <= 5 (chars)
- Token Name: length <= 30 (chars)
- Token Version: Should always be 1 (encoded in 1 byte)
- signature: length == 32 bytes

These informations will be encoded to be passed on the APDU command as:
```
token_uid[32] + len(symbol)[1] + symbol + len(name)[1] + name + version[1] + len(signature) + signature
```

This will give us a max of 205 bytes (using max length and when UTF-8 uses 4 bytes, 100 when using 1 byte)
Since an APDU command let's us pass up to 255 bytes of data we can always encode this command on 1 call.

### SEND_TOKEN_DATA

This command will work exactly like VERIFY_TOKEN_SIGNATURE but with a side effect of saving the token uid/symbol mapping on the memory if the signature verification was succesful.

This global variable of token uid/symbol has 38 bytes (32 of uid + 6 of symbol and null terminator) for each element with a total of 10 elements it will ocupy 380 bytes of RAM.
The 10 element limit is to impose a ceiling to how much RAM will be dedicated to this variable.
This imposes a limit on how many custom tokens can be used on transactions since all tokens should be loaded on memory when decoding the transaction.

### INVALIDATE_TOKEN_SIGNATURES

This new command will not receive any input, just ask the user confirmation of the action.

The screen should have the steps:
- Display warning symbol with this action name
- Display text to inform user that all token signatures will be voided
- Step with confirm action
- Step with reject action

This will be the only method that overwrites the permanent storage (after the first set).
If the user rejects simply return a reject code to the wallet.

### SIGN_TX

The current command receives the sighash_all data to sign and parses the output portion so the user can confirm the destination of the funds.

The token uid to symbol mapping should already be on the global context, sent with the SEND_TOKEN_DATA command.
If any token uid on the transaction is not present on the context the command should fail (HTR is excluded from this).

The output's `token_data` has a few functions:
- Identify authority outputs with the first bit
- The 7 other bits should be used as an unsigned integer
    - If the integer is 0 the output uses HTR
    - IF the integer is greater than 0, use integer minus 1 as index on the tokens array and the output is of that token

Instead of creating a token array for the sign_tx context, create a pointer array and populate with pointers to the global context token.
This way we can still use the token_data index on this pointer array and get a pointer to the token information previously confirmed with SEND_TOKEN_DATA.
The information will also be used to show the token symbol for the output value.

### Methods
[methods]: #methods

#### Signature

The signature will be a sha256 digest of the secret and token information.

```c
// message == secret + uid + symbol + name + version
cx_hash_sha256(message, message_len, signature, 32);
```

The cx_hash_sha256 gives a one-shot sha256 digest of the input data.
The verification will be a simple hash of the information and then verifying that it matches the received signature.

#### Persistent Storage

The secret used on both methods to sign and verify is saved and read from the permanent storage.
The memory address of this integer can be empty (it should be initialized with zero bytes).
If this case is found, we should generate the new integer and save ([using `nvm_write`](https://developers.ledger.com/docs/nano-app/persistent-storage/#types-of-memory)) except if the command is "INVALIDATE_TOKEN_SIGNATURES" (we cannot invalidate if there are no signatures to invalidade).

The secret will be 32 bytes of data generated with `cx_rng` (an sdk method that requests bytes from the ledger rng).
All app data are isolated from each other (including persistent storage), since no other app has access to it the secret is safe (This was confirmed by a Ledger engineer).

To make all cases posible we should implement methods to:
- `write`: Generate 32 bytes of random data with `cx_rng` and write to the persistent storage with `nvm_write`.
- `get`: Read the secret bytes, if all bytes are zero generate a new one and write, return the generated one.

### Wallet

The wallet should be able to have custom tokens registered even if they were not confirmed with the Ledger.
If the token is registered but not confirmed the user should be able to start the signing process for a single token.
The wallet should verify that the tokens it holds are valid on startup and before sending a transaction with custom tokens.

If the custom tokens are not signed by the time a transaction is being sent to the Ledger the wallet should confirm with the user if he wants to confirm the unknown tokens.

The method SEND_TOKEN_DATA should only be used before sending a transaction since it has a side effect of loading data on memory.

# Drawbacks
[drawbacks]: #drawbacks

- Saving the signature data on the wallet makes it so that if you change computers the user will have to confirm all tokens again.

- Having a single integer being the "salt generator" makes any changes on the integer a change on all signatures, so we cannot "untrust" a single signature.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

### Signature method

We considered using ecdsa signatures of the token data instead of a hash digest, this is because an ecdsa signature is linked with a keypair and it's the only thing that is secure enough to trust.
But after some considerations a sha256 with a big enough salt is just as secure as the salt storage.
After confirming with the Ledger team that the persistent storage is safe to store secrets (only our app has access to it) we changed to use a sha256 digest of the token data.

### Signature data location

The most secure location to store the signature data would be on the Ledger
But having too many erase/write cycles on the permanent storage will decrease the lifespan of the device.
The Ledger storage is also limited unlike the wallet, where we can add more metadata to make the UI/UX more friendly.

### Confirm the token during send_tx

The current command returning the signature is more reusable and developer friendly than adding optional sub-steps on the sign_tx depending on the data received.
We also enable confirming custom tokens outside of a sign_tx command, so the user don't have to send a token to someone (or himself) to start "trusting" the token.

### Restrict access to tokens

Another option of added security is to have a hard-coded public-key on the ledger and only verify token signatures using this public key.
This means we could control which tokens the ledger can (or cannot) make transactions with.

This removes the case of users being scammed or hacked since any token would need to pass an aproval process.
The token signature along with the information (token_uid symbol) would be public information that all wallets would have access to.

This option removes freedom from the user on the basis of security. This design opted for giving the user freedom for which tokens he wants to use.

### Other RNG methods on Ledger

We could generate an address outside of the public permitted path (but still on our bip44 path so no other app uses it).
The result would be deterministic and the same for all tokens, so we could add the token uid to the salt and make it unique for each custom token.
This method can be very secure since all information are saved inside Ledger and not accessible to other apps, but it depends on the path rules.
Changes to path rules could open the path segment we use to generate "random bytes" to the address command, making the salt knowable from outside the device.
The method to choose the path depends on what data we decide to use, a simple increasing 32bit integer saved on Ledger permanent storage could be open to attacks (since a reset could be made and it would reset the "count").
So there is a lib for RNG under `cx.h` with many useful methods like `cx_rng_u32`, `cx_rng`, `cx_rng_u32_range`, `cx_rng_rfc6979` (most use `cx_rng_no_throw` underneath).

### calloc/malloc for token uid symbol global

These are valid methods to reserve memory for the token uid/symbol on RAM, but there are many issues with this approach on a limited RAM space like Ledger (total 4K+512 bytes).
To avoid the issues of memory leak and overflow we prefer not to use this, we don't lose too much since the 10 token limit we imposed as an alternative solves pretty much all use-cases.

# Unresolved issues

This design does not contemplate:
- Custom token creation
- mint/melt operations
- Any transaction with an authority output

# Task Breakdown
[task-breakdown]: #task-breakdown

### Ledger

- [ ] Create the methods to sign/verify a signature defined in [methods](#methods)
- [ ] Create the methods to read/write on persistent storage (test the init of global persistent variables described [here](https://developers.ledger.com/docs/nano-app/persistent-storage/#types-of-memory))
- [ ] Create the INVALIDATE_TOKEN_SIGNATURES command
- [ ] Create the SIGN_TOKEN_DATA command
- [ ] Create the VERIFY_TOKEN_SIGNATURE command
- [ ] Create the SEND_TOKEN_DATA command
- [ ] Make changes to the SIGN_TX command (this should be heavily tested since it affects the current functionality of the app)
    - [ ] For each output, check that the token is present on the token_uid/symbol mapping made with SEND_TOKEN_DATA

### Wallet

- [ ] Create the new command handlers
- [ ] Create a button to delete all tokens (the signatures, not the actual funds) and call INVALIDATE_TOKEN_SIGNATURES
- [ ] Create a button to sign a known token that is still unsigned
- [ ] Call SEND_TOKEN_DATA whenever a hardware wallet tries to register a token (via token string or adding from unknown tokens)
- [ ] Before sending request to sign a tx, call SEND_TOKEN_DATA for all, then request SIGN_TX (user should be alerted!)
- [ ] Before INVALIDATE_TOKEN_SIGNATURES a prompt should be shown to the user with the consequences.
- [ ] After INVALIDATE_TOKEN_SIGNATURES a prompt should be shown to the user.
    - [ ] The prompt should offer to start a SEND_TOKEN_DATA for all tokens on the wallet.
    - [ ] The prompt should offer to delete all registered tokens (if not accepted only delete signature information)
    - [ ] Maybe include a dropdown with all tokens to optionally delete or issue a SEND_TOKEN_DATA on the steps above

**Comment by pedroferreira1** [comment]
I like the approach of having a single token verification which is more complex and detailed and then you can send many transactions using this token with simple symbol verification.

I would like to see more details on how the process will be. The user will register the token in the wallet, then the user verifies the token on ledger, then it can use it to send a transaction?

Seem like reasonable, we just need to discuss the UI/UX in the wallet and on ledger.

> Saving the signature data on the wallet makes it so that if you change computers the user will have to confirm all tokens again.

I don't see it as a big problem, to be honest.

**Comment by pedroferreira1** [comment]
The low level and the task breakdown seems really good to me. I have one question only, for what I understand we will use the private key from the hd derivation path where the last index will be the `uint32` saved in the permanent memory, right?

We can't support 2ˆ32 ints in the derivation path, we have 2ˆ31 as a limit I guess. Can we limit this r andom generated number to use at most 31 bits? Or am misunderstanding something?

**Comment by r4mmer** [comment]
> We can't support 2ˆ32 ints in the derivation path, we have 2ˆ31 as a limit I guess. Can we limit this r andom generated number to use at most 31 bits? Or am misunderstanding something?

While there is a method to generate a random 32bit integer in a given range, we do have 2^32 possibilities.
The first half being normal indices and the later half (greater then 2^31) are hardened indices. So a `i'` on the path would be the index `i+2^31`.

Since the derivation still generates a valid private/public key pair when using hardened indices I don't think there is any problem with generating a number in this space.
The path chosen was completely arbitrary as explained on the "Rationale and alternatives" section, so we have the freedom to change this path without affecting the safety of any signature.
The only thing we should avoid is having a static path.

**Comment by pedroferreira1** [comment]
Sound great. The design is approved for me

**Comment by obiyankenobi** [comment]
Did we research how this is done with tokens on ethereum? From what I remember, there's no such check on screen. What are their security measures?

> If the signature information on the wallet is leaked it could be used as an attack vector and the Ledger would have no way of knowing that the user has "untrusted" the token.

Could you clarify this? As far as I understand, this will be signed with a private key on Ledger. What can be leaked?

**Comment by msbrogli** [comment]
Here are my comments:

1) timestamp[14]: Why a string with 14 bytes instead of an integer number? What timezone will be used? UTC?

2) Can the ledger verify the timezone?

3) Will we use one salt per registered token? Or a global salt for all of them?

4) How much memory is available in the permanent storage?

5) As far as I understood, the tx_send method will fail in case an unconfirmed token is received. Is that right? What error code will be returned?

6) I don't have full understanding of the changes in the protocol.

7) I feel we can register the token in the Ledger when the token is registered in the wallet. What do you think?

8) It seems Hathor Wallet will have to manage which tokens have already been successfully registered in Ledger. In case of an error received during a send_tx, it should mark the token as not registered. Does it make sense?

9) What happens when a new wallet connects to a Trezor previously used by another wallet?

10) What happens when a new Trezor connects to a Wallet previously used by another Trezor?

**Comment by r4mmer** [comment]
> 1. timestamp[14]: Why a string with 14 bytes instead of an integer number? What timezone will be used? UTC?

This is a very optional input, it's just to add some information for the user to confirm that this request is being made now and to differentiate 2 signature requests (even with the same internal `i` the signature will be different).

> 2. Can the ledger verify the timezone?

I tried some c libs for dealing with timestamps but they don't compile. So i tried to make it as easy as posible to display useful information for the user while also adding some information bytes from the wallet.

> 3. Will we use one salt per registered token? Or a global salt for all of them?

The salt is generated from the internal integer and the token uid, so each token will have it's own salt.

> 4. How much memory is available in the permanent storage?

The permanent storage is the same storage where the app data is stored. The global variables meant for permanent storage will be saved alongside the app data on the ledger, so any memory used must be "reserved".
For instance, we could make an array of 512 entries of 4 bytes each, but it will add 2KB to the app size.

> 5. As far as I understood, the tx_send method will fail in case an unconfirmed token is received. Is that right? What error code will be returned?

Yes, it will fail and we will create a new SW code for this error (It will most likely be `0xB00A` since the last one is `0xB009`).

> 6. I don't have full understanding of the changes in the protocol.

Other than adding 2 new commands the only change will be on the sign_tx where we will send a list of token data and signatures.

> 7. I feel we can register the token in the Ledger when the token is registered in the wallet. What do you think?

This depends on a heavy usage of the persistent storage. I could test this possibility and discuss the results, what do you think?

> 8. It seems Hathor Wallet will have to manage which tokens have already been successfully registered in Ledger. In case of an error received during a send_tx, it should mark the token as not registered. Does it make sense?

Yes.

> 9. What happens when a new wallet connects to a Trezor previously used by another wallet?

It has to register any custom tokens it wants to use.

> What happens when a new Trezor connects to a Wallet previously used by another Trezor?

All signatures will fail, and then any attempt to sign a transaction will need to re-register the token being sent.

**Comment by obiyankenobi** [comment]
Could we also add in this design to display the token name in a separate Ledger step, including HTR? This way, we have more space on the amount screen and can add the `,` separator to facilitate reading the number,.

**Comment by jansegre** [comment]
> What happens when a new wallet connects to a Trezor previously used by another wallet?
> What happens when a new Trezor connects to a Wallet previously used by another Trezor?

Shouldn't it say _Ledger_ instead of _Trezor_ or did I miss something? @r4mmer

**Comment by obiyankenobi** [comment]
> On the signature command the token symbol and name are separate, and on the sign_tx command only the symbol is shown, we could add the name to the output confirmation but I think it will be too much information.

Sorry, I meant the tokens symbol during the sign_tx process.

**Comment by r4mmer** [comment]
> Sorry, I meant the tokens symbol during the sign_tx process.

Well I think we should include this change, but it would be more beneficial to have a wider discussion on UI/UX with ledger.

**Comment by r4mmer** [comment]
> Shouldn't it say Ledger instead of Trezor or did I miss something? @r4mmer

I copy/pasted @msbrogli question, I think he meant ledger as well.

**Comment by msbrogli** [comment]
I'm good with the current decisions. But it is still unclear what we expect from the wallets. Here are my thoughts:

1) Add an explanation that wallets can fully load from the xpub. In other words, wallets can load custom tokens even for non-registered tokens in Ledger.

2) Add a command to test a signature. It might be useful for wallets to check whether they signatures are still valid.

3) Let's say one has 200 different tokens. I understood that one will have to confirm each of the 200 tokens. I know each wallet can have a different implementation but what would be our recommended policy? Should they register the token in  the Ledger at the registration time in the wallet? Or should they postpone to register in the Ledger just when one tries to send a transaction? I feel we should add notes about this case in the design because it has impact in the user experience.

**Comment by r4mmer** [comment]
> Let's say one has 200 different tokens. I understood that one will have to confirm each of the 200 tokens. I know each wallet can have a different implementation but what would be our recommended policy? 

A transaction can have up to 255 tokens, but the ledger wouldn't be able to hold this many information on memory. So I split the verify token signature and send token data methods, which operates the exact same way but send token data saves the token information on memory to be used by the sign_tx command.

This memory concern is why I believe the wallet should verify tokens signatures on startup and only use the send command when trying to sign a transaction. The problem with this approach is that verifying 200 signatures may take a while, so this process should be done in parallel after startup.

**Comment by obiyankenobi** [comment]
> Did we research how this is done with tokens on ethereum? From what I remember, there's no such check on screen. What are their security measures?
> 
> > If the signature information on the wallet is leaked it could be used as an attack vector and the Ledger would have no way of knowing that the user has "untrusted" the token.
> 
> Could you clarify this? As far as I understand, this will be signed with a private key on Ledger. What can be leaked?

Can you answer these, please?

**Comment by r4mmer** [comment]
> Did we research how this is done with tokens on ethereum? From what I remember, there's no such check on screen. What are their security measures?

Yes, this research was done, ethereum is transitioning between 2 methods, on both the user has to send the token information before sending the transaction (like what we are doing). The "old" method is to have all permitted tokens hardcoded on the app and the new method is to have only a public key hardcoded (not from the user) and the wallet has to provide an ecdsa signature of the token data to verify with the hardcoded public key, this signature I believe is public information and only signed (i.e. trusted) tokens can be used on ledger but the user has no say on this. Our way leaves this decision on which tokens to trust to the user.

> Could you clarify this? As far as I understand, this will be signed with a private key on Ledger. What can be leaked?

The signature will be saved on the wallet, the information on the token is public so someone with access to the wallet storage could get the signature and save it. If we didn't have the salt (only known by ledger) the signature would be valid forever, making untrusting a token impossible.

**Comment by pedroferreira1** [comment]
>  Call SEND_TOKEN_DATA whenever a hardware wallet tries to register a token (via token string or adding from unknown tokens)

Should we make it required, or optional so the user can get the signature from ledger later?

> Before sending request to sign a tx, call SEND_TOKEN_DATA for all, then request SIGN_TX (user should be alerted!)

We may be sending a tx with a token without signatures, right? So we must assume that we could have another step to verify the tokens and get its signature before sending token data, right?

**Comment by r4mmer** [comment]
> Should we make it required, or optional so the user can get the signature from ledger later?

I believe we should make it optional, but have a checkbox for the user to confirm that he wants to register with the ledger.
Also, I changed the name of this method and forgot to update the task breakdown, it should be SIGN_TOKEN_DATA.

> We may be sending a tx with a token without signatures, right? So we must assume that we could have another step to verify the tokens and get its signature before sending token data, right?

If a token without signature is on the transaction the user is trying to send, the wallet should inform the user that he has to verify the token and start the SIGN_TOKEN_DATA command.
By using the SEND_TOKEN_DATA command on all tokens before sending the transaction the wallet also receives confirmation that all token signatures (on the tx) are valid. If any are not valid, the user should be informed and the wallet should start the SIGN_TOKEN_DATA command.

**Comment by andreabadesso** [comment]
I've just read the design, discussion, and decisions that were made and the final revision is looking great to me

This design is approved

**Comment by r4mmer** [comment]
During the implementation of the token context (which holds the verified tokens for a sign_tx command) I encountered a problem, to hold a token on memory we allocate 205 bytes, we can shorten this to 70 if we only allow ascii characters.

Using 205 bytes on memory for each token means to hold information on 5 tokens would make 1KB of memory on the global context, which is not dynamically allocated, so this 1KB would be reserved for all operations.

This means that we should have a ceiling of tokens we can use on any given transaction. I'm inclined to limit the size of the token by allowing only ascii characters and having another token type with only critical data (i.e. uid and symbol) and having a limit of 10 custom tokens, this reserves 370 bytes on memory.

I'm not considering using a `calloc`/`malloc` approach since it would not make much of a difference and would open the device to a memory allocation past it's limit which is very undesirable.

This does NOT mean that the user can only trust 10 tokens, he can have hundreds trusted, but only use 10 at a time on any given tx.

**Comment by andreabadesso** [comment]
Limiting it to a maximum of 10 tokens per transaction sounds very reasonable for the regular user

Transactions with that amount of tokens are probably not being generated by manually using the wallet (and the connected ledger)

**Comment by msbrogli** [comment]
I also agree it is reasonable to put a cap of 10 tokens per transaction.

**Comment by pedroferreira1** [comment]
Yeah, I agree that it's good enough. We just need to validate this in the wallet and add a documentation about all restrictions we are adding.

When we release the text with the announcement of this feature would be good to detail everything that is not being supported.

**Comment by jansegre** [comment]
> To avoid these cases we are only allowing ascii only symbols and names.

So our Ledger app will refuse transactions of tokens with a symbol that has a non-ascii character? I'm really not a fan of this. I'd prefer a confusing/weird/ugly UX/UI over not being able to transfer these tokens. If these were absolutely rare or nonexistent it could be reasonable, but there are several tokens with emojis on the symbol. I'd go with block characters as a first version, after that we can discuss counter measures or how UC can be improved.

> Token UID: lenth == 32 bytes

typo: length

> This will give us a max of 175 bytes (using max length and when UTF-8 uses 4 bytes, 70 when using 1 byte)

For a little while I was worried non-ascii chars could make this too big. But it seems fine. I'm even more pro not rejecting non-ascii chars now.

> If the custom tokens are not signed by the time a transaction is being sent to the Ledger the wallet should confirm with the user if he wants to confirm the unknown tokens.

I think I'm starting to get a better picture now. Maybe a few examples of the UX flow right at the beginning would make it much easier to follow up.

I may be convinced to refuse verifying tokens with non-ascii chars if the Ledger app will not refuse to sign transactions with unknown tokens.

Just to make it clear, the Desktop wallet is the one with the UX burden of suggesting the user verify unknown tokens, right? I understand Ledger app should have the minimal interactions but somehow I'm not 100% that's what was intended.

> The method SEND_TOKEN_DATA should only be used before sending a transaction since it has a side effect of loading data on memory.

I understood this. I'm really unsure about it. I guess the current limits are OK, but the design isn't great. I think it's fine for now, but IMO this should definitely generate a technical debt. Maybe the wallet could send verification "proofs" while the user accepts the inputs, even if it has to send info on the same toke multiple times, but anyway, I'm probably not familiar with Ledger app interactions and limitations to know better. A voice talk could really help me understand a lot better.

> Make changes to the SIGN_TX command (this should be heavily tested since it affects the current functionality of the app)

What would be the downsides of adding a SIGN_TOKEN_TX method instead?

I just realized I reviewed this without looking at the comments. I suspect many of my doubts could already have been answered. I'll post this as is but amend after reading all the comments.

**Comment by r4mmer** [comment]
>  I'd go with block characters as a first version, after that we can discuss counter measures or how UC can be improved.

This can be done, but for safety it was decided to not allow any ambiguous information. We can discuss this and it is relatively easy to replace utf-8 characters with blocks.

> Maybe the wallet could send verification "proofs" while the user accepts the inputs

There is no async interactions on ledger, and the idea was to not change the sign_tx command if the user is not sending custom tokens, this will make old version of wallets compatible with the new ledger version.

>  I think a similar option under some "advanced settings" would be extremely helpful. What do you think @r4mmer?

I think this is a good idea, but i'll have to look into the solana app to know what is being done.

**Comment by obiyankenobi** [comment]
I was thinking if people will have to register their custom tokens every time they use the hardware wallet. I ask that because we go back to the screen to chose between Hardware and Software wallet each time we disconnect the Ledger. I'm not sure if that clears the wallet data.
